### ISSUE TITLE
WebSocket is already in CLOSING or CLOSED state.

### ISSUE THREAD

**Issue by divyeshDev** [shamblett/mqtt_client#428]
I used this mqtt_client package.
Every time after 20-25 minutes WebSocket **WebSocket is already in CLOSING or CLOSED state.** gives an error.

Any suggestions on this issue? How to fix this problem?

OS : MacOs
Broswer : Chrome, safari 

![Screenshot 2022-12-15 at 10 12 02 AM](https://user-images.githubusercontent.com/104503242/207774243-81308f76-6cf8-4ac2-b618-eefd653c3250.png)

**Comment by shamblett** [comment]
OK, I'll need a client log to see whats happening here although this particular problem hasn't been reported before.

**Comment by divyeshDev** [comment]
Hello, @shamblett 

please look client logs:

![Screenshot 2022-12-28 at 1 30 37 PM](https://user-images.githubusercontent.com/104503242/209779422-a76f983e-fa97-4499-b03b-816e270feb07.png)

**Comment by shamblett** [comment]
The failure is in the connectAuto method so you must have lost your connection and triggered auto reconnect. It should stay in auto reconnect until it connects, does it eventually connect if you leave it long enough? I'll check the code to make sure we are recycling the socket.

**Comment by divyeshDev** [comment]
@shamblett For more than 30 minutes we do nothing on the web. Then we receive WebSocket error.

**Comment by shamblett** [comment]
Yes becsuse the network has disconnected you and cos u have autoreconnect set the client tries to reconnect, this is where the error appears. 

If you do nothing and let the client just run dors it eventually re connect(assuming your network comes back of course) ?

Its also possible your broker has gone down but this is rarer.

**Comment by divyeshDev** [comment]
No, I didn't turn off the network.
the network is continuously on my pc.

**Comment by shamblett** [comment]
Ok, in that case when the autoreconnect sequence starts can you still ping your broker?

**Comment by divyeshDev** [comment]
no, not.
@shamblett please see attached photo

![Screenshot 2022-12-28 at 4 33 51 PM](https://user-images.githubusercontent.com/104503242/209802429-b65987c0-b884-40d8-be5e-3d524a6d5185.png)

**Comment by shamblett** [comment]
No sorry I didnt mean from the client, the client is not connected so it will  not send keep alive pings I meant the standar networking ping command at the cli or any other way u may want to verify ur broker is still reachable.

**Comment by MurtuzaSrashtaSoft** [comment]
Client is already reachable. how we tell when i do browser refresh button in flutter hole application reinitialised and at that time new mqtt instance is created and that is working fine.

So broker is all time reachable and it's aws iot broker.

**Comment by shamblett** [comment]
OK thanks for checking, what we need to know now is why auto reconnect is being triggered in the first place, i.e. why we have(or think we have) lost connection to the broker in the first place.

 I need the client log from just above where autoreconnect is triggered to see why. the log you have supplied shows the client when its already in autoreconnect, this is too late.

**Comment by divyeshDev** [comment]
hello, @shamblett please look

2023-01-04 14:56:02.446 -- MqttBrowserWsConnection::connect - websocket is open
2023-01-04 15:14:53.868 -- MqttBrowserConnection::_startListening - websocket is closed


![Screenshot 2023-01-04 at 3 55 42 PM](https://user-images.githubusercontent.com/104503242/210535031-4381b6e4-38a5-4cc3-a847-52159871565d.png)



is there any callback available that tells WebSocket goes into the reconnected(MqttBrowserWsConnection::connectAuto - WebSocket has errored) mode?

**Comment by shamblett** [comment]
The first two lines, show that at 14:56 you connected and at 15:14 the socket closed for some reason as you have spotted, so you have lost connection at the network level which is triggering auto reconnect, this should work however. There is an on Autoreconenct callback that is triggered when autoreconnect kicks in, there is also one when autoreconnect has completed, please see the API docs.

**Comment by MurtuzaSrashtaSoft** [comment]
ok. and how we can identifying why it's goes to reconnect. and why auto reconnect not able to connect successfully?

**Comment by shamblett** [comment]
Unfortunately the client cant tell you why it was disconnected as youvr seen from the logs the client just getsca socket event from the Dart runtime saying the socket has closed its not told why.

Autoreconnect should work however Ill do some testing to see if we have bug here for browser based connections.

**Comment by MurtuzaSrashtaSoft** [comment]
@shamblett 

and this is heppend only on web environment in release mode in debug mode it's working fine.

now i don't have a any why how to resolve this issue.if you have any suggestion kindly share with us.

**Comment by shamblett** [comment]
When you say 'working fine' do you mean in debug mode when the network disconnects and autoreconnect starts then the client does eventually re-connect or are you saying  that the client doesn't disconnect in the first place?

The client package has no concept of debug or release mode, its just Dart code so whatever is happening I doubt if the client can help you with it.

**Comment by divyeshDev** [comment]
Any update?

is there any way to resolve this still we have received this WebSocket error?

Is it possible that when MQTT disconnects, it can connect automatically? Is there any solution?

**Comment by divyeshDev** [comment]
@shamblett Any update?

is there any way to resolve this still we have received this WebSocket error?

Is it possible that when MQTT disconnects, it can connect automatically? Is there any solution?

**Comment by ashley-arta** [comment]
Hi @shamblett I'm facing the same issue on web too.
It's not possible to connect with the same client identifier again

![image](https://github.com/shamblett/mqtt_client/assets/114375528/15c12d7a-6a8e-4f0e-b407-96a4748e0cc0)

**Comment by ashley-arta** [comment]
sorry my errors were actually caused by the Backend 🙇 kubernetes were down

**Comment by uberchilly** [comment]
This happens on Chrome-based browsers after the tab is being inactive for some time, it some sort of optimization from Chrome, it doesn't happen on Mozilla and you cannot reproduce it (at least I think) in debug mode for some reason.
